applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py

```diff
@@ -15,65 +20,80 @@ def __init__(self, Model, settings ):
         default_parameters = KratosMultiphysics.Parameters( """
             {
                 "model_part_name":"PLEASE_CHOOSE_MODEL_PART_NAME",
-                "mesh_id": 0,
-                "inlet_phi": 1.0,
-                "velocity_infinity": [1.0,0.0,0]
-            }  """ );
-
-
+                "inlet_potential": 1.0,
+                "free_stream_velocity": [1.0,0.0,0]
+            }  """ )
         settings.ValidateAndAssignDefaults(default_parameters);
 
-        self.model_part = Model[settings["model_part_name"].GetString()]
-        self.velocity_infinity = KratosMultiphysics.Vector(3)#array('d', [1.0, 2.0, 3.14])#np.array([0,0,0])#np.zeros(3)#vector(3)
-        self.velocity_infinity[0] = settings["velocity_infinity"][0].GetDouble()
-        self.velocity_infinity[1] = settings["velocity_infinity"][1].GetDouble()
-        self.velocity_infinity[2] = settings["velocity_infinity"][2].GetDouble()
-        #self.density_infinity = settings["density_infinity"].GetDouble() #TODO: must read this from the properties
-        self.inlet_phi = settings["inlet_phi"].GetDouble()
-        self.model_part.ProcessInfo.SetValue(CompressiblePotentialFlowApplication.VELOCITY_INFINITY,self.velocity_infinity)
-
+        self.far_field_model_part = Model[settings["model_part_name"].GetString()]
+        self.free_stream_velocity = settings["free_stream_velocity"].GetVector()
+        self.inlet_potential_0 = settings["inlet_potential"].GetDouble()
+        self.far_field_model_part.ProcessInfo.SetValue(CPFApp.VELOCITY_INFINITY,self.free_stream_velocity)
 
+    def ExecuteInitializeSolutionStep(self):
+        self.Execute()
 
     def Execute(self):
-        #KratosMultiphysics.VariableUtils().SetVectorVar(CompressiblePotentialFlowApplication.VELOCITY_INFINITY, self.velocity_infinity, self.model_part.Conditions)
-        for cond in self.model_part.Conditions:
-            cond.SetValue(CompressiblePotentialFlowApplication.VELOCITY_INFINITY, self.velocity_infinity)
+        reference_inlet_node = self._FindFarthestUpstreamBoundaryNode()
+        self._AssignFarFieldBoundaryConditions(reference_inlet_node)
 
-        #select the first node
-        for node in self.model_part.Nodes:
-            node1 = node
+    def _FindFarthestUpstreamBoundaryNode(self):
+        # The farthes upstream boundary node is the node with smallest
+        # projection of its distante to a given reference node onto
+        # the free stream velocity.
+
+        # Select a random node in the mesh as reference (e.g. the first node)
+        for node in self.far_field_model_part.Nodes:
+            reference_node = node
```

Comment:
I believe the `reference_node` has no effect on finding the `reference_inlet_node`. So you can simplify this code further.

Reply:
You are completely right, I do not know why we are doing this. One could directly do:

```python
def _FindFarthestUpstreamBoundaryNode(self):
        # The farthes upstream boundary node is the node with smallest
        # projection of its position vector onto the free stream velocity.

        # Find the farthest upstream boundary node
        temporal_smallest_projection = 1e30
        for node in self.far_field_model_part.Nodes:
            # Projecting the node position vector onto the free stream velocity
            distance_projection = DotProduct(node, self.free_stream_velocity)

            if(distance_projection < temporal_smallest_projection):
                temporal_smallest_projection = distance_projection
                reference_inlet_node = node

        return reference_inlet_node
```